### ISSUE TITLE
[3.x] Performance confusion?

### ISSUE THREAD

**Issue by Dadaskis** [godotengine/godot#87892]
### Tested versions

Reproducible in: 3.5.3.stable., 3.6.beta4

### System information

Windows 11 - Godot v3.5.3.stable - GLES3 - Nvidia Geforce GT1030

### Issue description

I'm not very certain what kind of title does this issue need to have, as i'm not very certain what exactly do i have to report. Nearly ago, i've switched over Windows 11. That resulted in performance drop of certain games, including my own project in Godot, but, if in other games it wasn't too much of a drop, in this project it turned the game to unplayable state. For about several days i've been trying to invent some new ways to optimize the game, like addition of LODs, physics cache, better handling of shader compilation, and some other settings related to environmental quality of the game. Nothing much helps, maybe i get some little fps from that, but mainly that is still resulting in a huge problem.

I want to mention that i'm still a novice in this engine (including terms of understanding how does GPU rendering and physics work) and many of the reported issues were just my own oversight.
https://github.com/godotengine/godot/issues/87665
https://github.com/godotengine/godot/issues/81461
And i hope that this time, i did something wrong once again.

Hardware Specifications:
GPU: Nvidia Geforce GT1030
RAM: 10 GB
CPU: AMD Athlon(tm) II X2 250 Processor 3.00 GHz
OS: Windows 11 23H2

Probably i expect too much of how much this hardware can do in terms of such project, however it still seems to be a resolvable issue, i have some reasons to think so. Potentially i can even overestimate an actual possibilities of this engine, but i'm still very sure that this game isn't that huge as it seems and can be developed in a way i intended it to be.

Before you mention that docs have tutorials regarding optimization of Godot 3D games, i have to mention that i've taken a look on following pages:
https://docs.godotengine.org/en/3.5/tutorials/performance/index.html
https://docs.godotengine.org/en/3.5/tutorials/performance/general_optimization.html
https://docs.godotengine.org/en/3.5/tutorials/performance/using_servers.html
https://docs.godotengine.org/en/3.5/tutorials/performance/cpu_optimization.html
https://docs.godotengine.org/en/3.5/tutorials/performance/gpu_optimization.html
https://docs.godotengine.org/en/3.5/tutorials/performance/optimizing_3d_performance.html

Firstly, i'm going to mention the new settings...
![image](https://github.com/godotengine/godot/assets/37717199/9e363591-4c9e-4f84-b1a8-6a80bb6b17d9)

Depth Quality - Controls either if depth is using HDR or not. High enables HDR, low setting disables it. Usually results in performance boost.

Particle Quality - High quality of that setting keeps all particles in the scene, that is dust motes, smoke, etc. That setting also includes fake volumetric lights that heavily impact the overdraw. Low quality of that setting, as it follows, completely removes all of these or keeps certain that are essential for the scene.

Material Quality - Controls either if materials are using full PBR texture set or not. If set on high, it keeps the PBR, if set on low, it does following things to all locational materials:
- Switching diffuse mode over "lambert".
- Switching specular mode over "disabled".
- Disabling normal maps and removing shaderparams related to it.
- Disabling MRSAO/MRS maps (channel packing), also removing params.
- Disabling all other kinds of things that may affect roughness, specular and metallic channels.

As it follows, material quality setting should decrease an amount of `glUniform`, `glBindTexture` and `glActiveTexture` calls.

Post-Processing Quality - Affects either if custom post-processing is using higher-end shaders or not. That usually means less use of `textureLod` and `SCREEN_TEXTURE`.

Character Quality - Human models in this game have LOD variations. Normal variation takes up for something like 10000-15000 triangles, while LOD variation lowering that amount to 3000. If that setting is set to low, it is forcing the game to render LOD variation all the time.

So after doing all of that stuff, i wanted to check how does it affect the game with GPU profiler named "Nsight Graphics".

![image](https://github.com/godotengine/godot/assets/37717199/adbcb96b-0f1a-4e9a-952e-2155088e86eb)
![image](https://github.com/godotengine/godot/assets/37717199/17b97e68-709a-4127-b684-54152f90058a)

Top screenshot keeps all of the mentioned settings on "high" while the lower one keeps them on "low". As you can see, amount of GPU events is lowered, however, it doesn't really give that much of a performance boost in the given scenario. 

During GPU profiling i noticed 2 interesting warnings...

![image](https://github.com/godotengine/godot/assets/37717199/d1d8c600-2a88-40ae-be2b-6a871a8024f5)
![image](https://github.com/godotengine/godot/assets/37717199/029e63e0-c8e7-4035-82a1-83d3c332d53a)

I wonder what they really mean, but they shouldn't be too critical, right?

That general performance problem might be a reason to actually take a look over what would Godot's profiler say, and... That is problematic, too.

Excuse me for a bit laggy footage, but i'm going to send it and mark what exactly is going on in specific timelines.

https://github.com/godotengine/godot/assets/37717199/f6bdc6f5-d6e1-4031-9ff3-e668747400ad

0:00 - Starting the profiler, starting the game's main scene (game_start.scn)
0:08 - Waiting it to launch, so i did a little of speeding up
0:28 - Once the game is ready, switching over Godot Editor and going to Debugger/Profiler tab.
0:33 - Witnessing that profiler has nothing in it, despite the game starting with profiler on.
0:36 - Switching over game, clicking once so the window gets focus.
0:40 - Game freezes, profiler shows something.
0:48 - Waiting it to unfreeze.
1:00 - Still waiting it to unfreeze. I did a bit of speeding up, so, basically i had to wait a minute.
1:03 - Switching over profiler to see that it is reporting that the game is running. I have to notice that the list of script functions is quite weird. It is reporting certain functions as "anonymous" ones even though all scripts used in the scene have `class_name` assigned to them.
1:08 - Script functions list suddenly purges itself.

However, i have to mention that profiler has another kind of weird behavior.

https://github.com/godotengine/godot/assets/37717199/19b649b8-fed1-4743-84bd-0c46d10bd171

0:00 - Starting the profiler, starting the game's main scene (game_start.scn)
0:10 - Switching over Debugging/Profiler tab while the game is starting to make the profiler work right from the start.
0:15 - Seeing that profiler window is visible, but the game itself is still launching.
0:16 - The game is launching.
0:23 - Profiler starts to display data.
0:45 - Switching over Godot Editor to see what does it display in profiler tab.
0:48 - There are still functions that are displayed as if they belonged to scripts with no `class_name`.
0:50 - Pausing the game using "Pause Scene" button.
0:52 - Profiler pauses only after 2 seconds of work.
0:56 - Turning the pause off.
1:10 - Turning it on again to see how profiler behaves.
1:17 - After 7 seconds of work, profiler stops.

When i used Windows 7, profiler used to work in more intuitive way (like pausing right after you pause the scene), additionally it used to provide more reliable set of data to analyze. Now it is questionable, but i cannot tell what exactly is going wrong.

### Steps to reproduce

No steps to reproduce yet.

### Minimal reproduction project (MRP)

Discussion of the problem for now, maybe i'll make MRP to recreate mentioned problems, but i literally have no clue what is going on.

**Comment by Dadaskis** [comment]
Profiler works fine on Godot 3.6. beta 4. I've noticed certain performance hogs on CPU side, going to fix them.

**Comment by Dadaskis** [comment]
3.5.3.stable - got the profiler working, but there are some exceptions. Instead of starting the game with it, i'm just running the game first and start profiler when i need it. It works this way as it should

**Comment by Dadaskis** [comment]
![image](https://github.com/godotengine/godot/assets/37717199/c7924578-3359-4dfe-94fa-0f74f1543600)

Going to write my report regarding performance of this game, if that is going to be helpful for anyone.

As you understand, i really want it to run on AMD Athlon, but that is kinda derelict and not very powerful CPU for this day. I've been thinking of how to optimize the game and came to conclusion - if i wanted it to be that optimized, i had to add nothing in first place. So i've added a "Game Settings" menu that would allow yourself to modify this game in a way to get more CPU performance and it will also allow yourself to disable parts of the game that you may not really like for certain reasons

For now, from laggy 3-8 fps it's going up to 25-30 fps during active combat scenarios, and there is still plenty of possibilities to work with it

TL;DR; All problems i have for now come from very intensive CPU usage

**Comment by Dadaskis** [comment]
Going to close it as it doesn't really matter now.

If anyone else having a similar issue with profiler: Try running the game first and start profiler only when you really need it
